symengine/relationals.cpp

```diff
@@ -0,0 +1,236 @@
+#include <symengine/relationals.h>
+
+namespace SymEngine
+{
+
+Relational::Relational() : lhs_{zero}, rhs_{zero}
+{
+    SYMENGINE_ASSIGN_TYPEID();
+}
+
+Relational::Relational(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs)
+    : lhs_{lhs}, rhs_{rhs}
+{
+    SYMENGINE_ASSIGN_TYPEID();
+}
+
+hash_t Relational::__hash__() const
+{
+    hash_t seed = RELATIONAL;
+    hash_combine<Basic>(seed, *lhs_);
+    hash_combine<Basic>(seed, *rhs_);
+    return seed;
+}
+
+inline RCP<const Basic> &Relational::get_lhs() const
+{
+    return lhs_;
+}
+
+inline RCP<const Basic> &Relational::get_rhs() const
+{
+    return rhs_;
+}
+
+inline vec_basic Relational::get_args() const
+{
+    vec_basic v;
+    v.push_back(lhs_);
+    v.push_back(rhs_);
+    return v;
+}
+
+inline bool Relational::__eq__(const Basic &o) const
+{
+    if (is_a<Relational>(o)) {
+        const Relational &s = down_cast<const Relational &>(o);
+        return (eq(*lhs_, *(s.get_lhs())) and eq(*rhs_, *(s.get_rhs())));
+    }
+    return false;
+}
+
+inline int Relational::compare(const Basic &o) const
+{
+    SYMENGINE_ASSERT(is_a<Relational>(o));
+    const Relational &r = down_cast<const Relational &>(o);
+    int cmp = lhs_->__cmp__(*(r.get_lhs()));
+    if (cmp != 0)
+        return cmp;
+    cmp = rhs_->__cmp__(*(r.get_rhs()));
+    return cmp;
+}
+
+RCP<const Basic> Relational::canonical() const
```

Comment:
In SymEngine, we keep all of the objects in the canonical form. I would change the design to create a `Relational` only in the canonical form